// botfuel-dialog/build/adapters/messenger-adapter.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _get = function get(object, property, receiver) { if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Copyright (c) 2017 - present, Botfuel (https://www.botfuel.io).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

var rp = require('request-promise-native');
var logger = require('logtown')('MessengerAdapter');
var PostbackMessage = require('../messages/postback-message');
var UserImageMessage = require('../messages/user-image-message');
var UserTextMessage = require('../messages/user-text-message');
var WebAdapter = require('./web-adapter');

var FB_GRAPH_URL = 'https://graph.facebook.com/v2.6';

/**
 * Adapter for the Facebook Messenger messaging platform.
 * @extends WebAdapter
 */

var MessengerAdapter = function (_WebAdapter) {
  _inherits(MessengerAdapter, _WebAdapter);

  function MessengerAdapter() {
    _classCallCheck(this, MessengerAdapter);

    return _possibleConstructorReturn(this, (MessengerAdapter.__proto__ || Object.getPrototypeOf(MessengerAdapter)).apply(this, arguments));
  }

  _createClass(MessengerAdapter, [{
    key: 'createRoutes',

    /** @inheritDoc */
    value: function createRoutes(app) {
      var _this2 = this;

      logger.debug('createRoutes');
      _get(MessengerAdapter.prototype.__proto__ || Object.getPrototypeOf(MessengerAdapter.prototype), 'createRoutes', this).call(this, app);
      app.get('/webhook', function (req, res) {
        return _this2.validateWebhook(req, res);
      });
    }

    /**
     * Webhook used by Facebook Messenger to validate the bot.
     * @async
     * @private
     * @param {Object} req - the request object
     * @param {Object} res - the response object
     * @returns {Promise.<void>}
     */

  }, {
    key: 'validateWebhook',
    value: function () {
      var _ref = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(req, res) {
        return regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                logger.debug('validateWebhook');
                if (req.query['hub.mode'] === 'subscribe' && req.query['hub.verify_token'] === process.env.FB_VERIFY_TOKEN) {
                  logger.debug('validateWebhook: OK!');
                  res.status(200).send(req.query['hub.challenge']);
                } else {
                  logger.error('validateWebhook: KO!');
                  res.sendStatus(403);
                }

              case 2:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function validateWebhook(_x, _x2) {
        return _ref.apply(this, arguments);
      }

      return validateWebhook;
    }()

    /** @inheritDoc */

  }, {
    key: 'handleRequest',
    value: function () {
      var _ref2 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee3(req, res) {
        var _this3 = this;

        var _req$body, object, entry;

        return regeneratorRuntime.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                logger.debug('handleRequest', req.body);
                _req$body = req.body, object = _req$body.object, entry = _req$body.entry;

                if (object === 'page') {
                  entry.forEach(function (entryItem) {
                    entryItem.messaging.forEach(function () {
                      var _ref3 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee2(event) {
                        return regeneratorRuntime.wrap(function _callee2$(_context2) {
                          while (1) {
                            switch (_context2.prev = _context2.next) {
                              case 0:
                                _context2.next = 2;
                                return _this3.processEvent(event);

                              case 2:
                              case 'end':
                                return _context2.stop();
                            }
                          }
                        }, _callee2, _this3);
                      }));

                      return function (_x5) {
                        return _ref3.apply(this, arguments);
                      };
                    }());
                  });
                  res.sendStatus(200);
                }

              case 3:
              case 'end':
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function handleRequest(_x3, _x4) {
        return _ref2.apply(this, arguments);
      }

      return handleRequest;
    }()

    /**
     * Processes a received event (message, postback, ...).
     * @async
     * @param {Object} event - the messenger event
     * @returns {Promise.<void>}
     */

  }, {
    key: 'processEvent',
    value: function () {
      var _ref4 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee4(event) {
        var sender, message, postback, userMessage, text, attachments, _JSON$parse, dialog, entities;

        return regeneratorRuntime.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                logger.debug('processEvent', JSON.stringify(event));
                sender = event.sender, message = event.message, postback = event.postback;
                userMessage = null;

                if (message) {
                  text = message.text, attachments = message.attachments;
                  // user send attachments

                  if (attachments && attachments[0].type === 'image') {
                    userMessage = new UserImageMessage(attachments[0].payload);
                  } else {
                    userMessage = new UserTextMessage(text);
                  }
                } else if (postback) {
                  _JSON$parse = JSON.parse(postback.payload), dialog = _JSON$parse.dialog, entities = _JSON$parse.entities;

                  userMessage = new PostbackMessage(dialog, entities);
                }
                _context4.next = 6;
                return this.handleMessage(userMessage.toJson(sender.id));

              case 6:
              case 'end':
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function processEvent(_x6) {
        return _ref4.apply(this, arguments);
      }

      return processEvent;
    }()

    /** @inheritDoc */

  }, {
    key: 'initUserIfNecessary',
    value: function () {
      var _ref5 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee5(userId) {
        return regeneratorRuntime.wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                _context5.next = 2;
                return _get(MessengerAdapter.prototype.__proto__ || Object.getPrototypeOf(MessengerAdapter.prototype), 'initUserIfNecessary', this).call(this, userId);

              case 2:
                _context5.next = 4;
                return this.updateUserProfile(userId);

              case 4:
              case 'end':
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function initUserIfNecessary(_x7) {
        return _ref5.apply(this, arguments);
      }

      return initUserIfNecessary;
    }()

    /** @inheritDoc */

  }, {
    key: 'getUri',
    value: function getUri() {
      return FB_GRAPH_URL + '/me/messages';
    }

    /** @inheritDoc */

  }, {
    key: 'getQs',
    value: function getQs() {
      return {
        access_token: process.env.FB_PAGE_ACCESS_TOKEN
      };
    }

    /** @inheritDoc */

  }, {
    key: 'getBody',
    value: function getBody(botMessage) {
      var message = this.adapt(botMessage);
      return {
        recipient: {
          id: botMessage.user
        },
        message: message
      };
    }

    /**
     * @private
     * @param {Object} payload - the payload
     * @returns {Object} the text
     */

  }, {
    key: 'adaptText',
    value: function adaptText(payload) {
      return {
        text: payload.value
      };
    }

    /**
     * @private
     * @param {Object} payload - the payload
     * @returns {Object} the quickreplies
     */

  }, {
    key: 'adaptQuickreplies',
    value: function adaptQuickreplies(payload) {
      return {
        text: payload.options.text,
        quick_replies: payload.value.map(function (qr) {
          return {
            content_type: 'text',
            title: qr,
            payload: qr
          };
        })
      };
    }

    /**
     * @private
     * @param {Object} payload - the payload
     * @returns {Object} the image
     */

  }, {
    key: 'adaptImage',
    value: function adaptImage(payload) {
      return {
        attachment: {
          type: 'image',
          payload: {
            url: payload.value
          }
        }
      };
    }

    /**
     * @private
     * @param {Object} payload - the payload
     * @returns {Object} the actions
     */

  }, {
    key: 'adaptActions',
    value: function adaptActions(payload) {
      logger.debug('adaptActions', payload);
      return {
        attachment: {
          type: 'template',
          payload: {
            template_type: 'button',
            text: payload.options.text,
            buttons: payload.value.map(MessengerAdapter.adaptAction)
          }
        }
      };
    }

    /**
     * @private
     * @param {Object} payload - the payload
     * @returns {Object} the cards
     */

  }, {
    key: 'adaptCards',
    value: function adaptCards(payload) {
      logger.debug('adaptCards', payload);
      var elements = payload.value.map(function (card) {
        var buttons = card.buttons.map(MessengerAdapter.adaptAction);
        return Object.assign(card, { buttons: buttons });
      });
      return {
        attachment: {
          type: 'template',
          payload: {
            template_type: 'generic',
            elements: elements
          }
        }
      };
    }

    /**
     * Adapts payload.
     * @private
     * @param {Object} botMessage - the bot message
     * @returns {Object} the adapted message
     */

  }, {
    key: 'adapt',
    value: function adapt(botMessage) {
      logger.debug('adapt', botMessage);
      var payload = botMessage.payload;
      switch (botMessage.type) {
        case 'text':
          return this.adaptText(payload);
        case 'quickreplies':
          return this.adaptQuickreplies(payload);
        case 'image':
          return this.adaptImage(payload);
        case 'actions':
          return this.adaptActions(payload);
        case 'cards':
          return this.adaptCards(payload);
        default:
          return null;
      }
    }

    /**
     * @private
     * @static
     * @param {Object} action - the action object
     * @returns {Object|null} the adapted action or null if action type is not valid
     */

  }, {
    key: 'updateUserProfile',


    /**
     * Get user profile informations and store it into the brain
     * @param {String} userId - the user id
     * @returns {void}
     */
    value: function () {
      var _ref6 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee6(userId) {
        var userProfile, res, profile;
        return regeneratorRuntime.wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                logger.debug('updateUserProfile', userId);
                // check for user profile informations
                _context6.next = 3;
                return this.bot.brain.userGet(userId, 'profile');

              case 3:
                userProfile = _context6.sent;

                if (!(!userProfile || !Object.keys(userProfile).length)) {
                  _context6.next = 18;
                  break;
                }

                _context6.prev = 5;
                _context6.next = 8;
                return rp({
                  method: 'GET',
                  json: true,
                  uri: FB_GRAPH_URL + '/' + userId,
                  qs: {
                    fields: 'first_name,last_name,gender',
                    access_token: process.env.FB_PAGE_ACCESS_TOKEN
                  }
                }).promise();

              case 8:
                res = _context6.sent;
                profile = {
                  firstName: res.first_name,
                  lastName: res.last_name,
                  gender: res.gender
                };
                _context6.next = 12;
                return this.bot.brain.userSet(userId, 'profile', profile);

              case 12:
                logger.debug('updateUserProfile: user profile updated with', profile);
                _context6.next = 18;
                break;

              case 15:
                _context6.prev = 15;
                _context6.t0 = _context6['catch'](5);

                logger.error('updateUserProfile: error', _context6.t0.message || _context6.t0.error || _context6.t0);

              case 18:
              case 'end':
                return _context6.stop();
            }
          }
        }, _callee6, this, [[5, 15]]);
      }));

      function updateUserProfile(_x8) {
        return _ref6.apply(this, arguments);
      }

      return updateUserProfile;
    }()
  }], [{
    key: 'adaptAction',
    value: function adaptAction(action) {
      logger.debug('adaptAction', action);
      switch (action.type) {
        case 'postback':
          return {
            type: 'postback',
            title: action.text,
            payload: JSON.stringify(action.value)
          };
        case 'link':
          return {
            type: 'web_url',
            title: action.text,
            url: action.value
          };
        default:
          return null;
      }
    }
  }]);

  return MessengerAdapter;
}(WebAdapter);

module.exports = MessengerAdapter;
