// botfuel-qna-sdk/build/api/v1/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _apiResource = require('../api-resource');

var _apiResource2 = _interopRequireDefault(_apiResource);

var _config = require('../../config');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var QnAV1 = function (_ApiResource) {
  _inherits(QnAV1, _ApiResource);

  function QnAV1(_ref) {
    var appId = _ref.appId,
        appKey = _ref.appKey;

    _classCallCheck(this, QnAV1);

    var _this = _possibleConstructorReturn(this, (QnAV1.__proto__ || Object.getPrototypeOf(QnAV1)).call(this));

    _this.headers = {
      'App-Id': appId,
      'App-Key': appKey
    };
    _this.API = _config.QNA_API_V1;
    return _this;
  }

  // ## Bots CRUD ## //

  _createClass(QnAV1, [{
    key: 'getBots',
    value: function getBots() {
      return this.curriedRp('/bots', {
        method: 'GET'
      });
    }
  }, {
    key: 'getBot',
    value: function getBot(_ref2) {
      var botId = _ref2.botId;

      return this.curriedRp('/bots/' + botId, {
        method: 'GET'
      });
    }
  }, {
    key: 'createBot',
    value: function createBot(_ref3) {
      var botId = _ref3.botId,
          name = _ref3.name,
          language = _ref3.language,
          response_nb = _ref3.response_nb;

      return this.curriedRp('/bots', {
        method: 'POST',
        body: {
          bot_id: botId,
          name: name,
          language: language,
          response_nb: response_nb
        }
      });
    }
  }, {
    key: 'updateBot',
    value: function updateBot(_ref4) {
      var name = _ref4.name,
          language = _ref4.language,
          response_nb = _ref4.response_nb;

      return this.curriedRp('/bots', {
        method: 'PUT',
        body: {
          name: name,
          language: language,
          response_nb: response_nb
        }
      });
    }
  }, {
    key: 'deleteBot',
    value: function deleteBot() {
      return this.curriedRp('/bots', {
        method: 'DELETE'
      });
    }

    // ## QnAs CRUD ## //

  }, {
    key: 'getQnAs',
    value: function getQnAs(_ref5) {
      var page = _ref5.page;

      return this.curriedRp('/bots/qnas', {
        method: 'GET',
        qs: {
          page: page
        }
      });
    }
  }, {
    key: 'getQnA',
    value: function getQnA(_ref6) {
      var qnaId = _ref6.qnaId;

      return this.curriedRp('/bots/qnas/' + qnaId, {
        method: 'GET'
      });
    }
  }, {
    key: 'createQnA',
    value: function createQnA(_ref7) {
      var name = _ref7.name,
          questions = _ref7.questions;

      return this.curriedRp('/bots/qnas', {
        method: 'POST',
        body: {
          name: name,
          questions: questions
        }
      });
    }
  }, {
    key: 'updateQnA',
    value: function updateQnA(_ref8) {
      var qnaId = _ref8.qnaId,
          name = _ref8.name,
          questions = _ref8.questions;

      return this.curriedRp('/bots/qnas/' + qnaId, {
        method: 'PUT',
        body: {
          name: name,
          questions: questions
        }
      });
    }
  }, {
    key: 'deleteQnA',
    value: function deleteQnA(_ref9) {
      var qnaId = _ref9.qnaId;

      return this.curriedRp('/bots/qnas/' + qnaId, {
        method: 'DELETE'
      });
    }

    // ## Corpus CRUD ## //

  }, {
    key: 'getBotCorpus',
    value: function getBotCorpus(_ref10) {
      var corpusId = _ref10.corpusId;

      return this.curriedRp('/bots/corpus/' + corpusId, {
        method: 'GET'
      });
    }
  }, {
    key: 'getBotCorpora',
    value: function getBotCorpora(_ref11) {
      var page = _ref11.page,
          size = _ref11.size;

      return this.curriedRp('/bots/corpus', {
        method: 'GET',
        qs: { page: page, size: size }
      });
    }
  }, {
    key: 'createBotCorpus',
    value: function createBotCorpus(_ref12) {
      var name = _ref12.name,
          words = _ref12.words;

      return this.curriedRp('/bots/corpus', {
        method: 'POST',
        body: {
          name: name,
          words: words
        }
      });
    }
  }, {
    key: 'updateBotCorpus',
    value: function updateBotCorpus(_ref13) {
      var corpusId = _ref13.corpusId,
          name = _ref13.name,
          words = _ref13.words;

      return this.curriedRp('/bots/corpus/' + corpusId, {
        method: 'PUT',
        body: {
          name: name,
          words: words
        }
      });
    }
  }, {
    key: 'deleteBotCorpus',
    value: function deleteBotCorpus(_ref14) {
      var corpusId = _ref14.corpusId;

      return this.curriedRp('/bots/corpus/' + corpusId, {
        method: 'DELETE'
      });
    }

    // ## Test, training, stats, correlations, model, classification ## //

  }, {
    key: 'testBot',
    value: function testBot(_ref15) {
      var threshold = _ref15.threshold;

      return this.curriedRp('/bots/test', {
        method: 'GET',
        qs: {
          threshold: threshold
        }
      });
    }
  }, {
    key: 'trainBot',
    value: function trainBot() {
      return this.curriedRp('/bots/train', {
        method: 'GET'
      });
    }
  }, {
    key: 'getBotStats',
    value: function getBotStats() {
      return this.curriedRp('/bots/stats', {
        method: 'GET'
      });
    }
  }, {
    key: 'getBotCorrelations',
    value: function getBotCorrelations() {
      return this.curriedRp('/bots/correlations', {
        method: 'GET'
      });
    }
  }, {
    key: 'getBotCorrelation',
    value: function getBotCorrelation(_ref16) {
      var qnaId = _ref16.qnaId;

      return this.curriedRp('/bots/correlations/' + qnaId, {
        method: 'GET'
      });
    }
  }, {
    key: 'getBotModel',
    value: function getBotModel() {
      return this.curriedRp('/bots/model', {
        method: 'GET'
      });
    }
  }, {
    key: 'getMatchingQnas',
    value: function getMatchingQnas(_ref17) {
      var sentence = _ref17.sentence;

      return this.curriedRp('/bots/classify', {
        method: 'POST',
        body: {
          sentence: sentence
        }
      });
    }
  }]);

  return QnAV1;
}(_apiResource2.default);

exports.default = QnAV1;
